warc_metadata_sidecar.py

```diff
@@ -0,0 +1,261 @@
+            if not record_count:  # and record.rec_type == 'warcinfo':
+                rec_type = record.rec_type
+                warc_info = create_warcinfo_payload(payload, operator, publisher,
+                                                    new_file, version)
+                # Write warcinfo record into sidecar.
+                writer = WARCWriter(output, gzip=False)  # TODO: gzip will equal True
+                warcinfo_record = writer.create_warcinfo_record(meta_file, warc_info)
+                writer.write_record(warcinfo_record)
+                record_count += 1
+            if record.rec_type not in ['response', 'resource']:
+                continue
+            payload.seek(0)
+            # The payload is how we find the important info. Skip record if empty.
+            if not len(payload.read()):
+                continue
+            if 'text/dns' in record.rec_headers.get_header('Content-Type'):
+                continue
+            record_count += 1
+            record_date = record.rec_headers.get_header('WARC-Date')
+            warcinfo_id = record.rec_headers.get_header('WARC-Warcinfo-ID')
+            warcrecord_id = record.rec_headers.get_header('WARC-Record-ID')
+            # Define specific warc_headers to include in sidecar.
+            warc_dict = {'WARC-Date': record_date, 'WARC-Concurrent-ID': warcrecord_id}
+            if warcinfo_id:
+                warc_dict['WARC-Warcinfo-ID'] = warcinfo_id
+
+            print(url)
+            payload.seek(0)
+            find_mime_and_puid(fido, payload, url)
+
+            if fido.mime:
+                text_format_mimes = re.compile(r'(text|html|xml)')
+                if text_format_mimes.search(fido.mime):
+                    payload.seek(0)
+                    bytes_payload = payload.read()
+                    result_dict = find_character_set(bytes_payload)
+                    lang_cld = find_language(bytes_payload)
+                    text_mime += 1
+                    if not lang_cld:
+                        string_payload = '{0} {1}\n{2} {3}\n{4} {5}'.format(
+                            mime_title, fido.mime,
+                            puid_title, fido.puid,
+                            charset_title, result_dict
+                            )
+                    else:
+                        string_payload = '{0} {1}\n{2} {3}\n{4} {5}\n{6} {7}'.format(
+                            mime_title, fido.mime,
+                            puid_title, fido.puid,
+                            charset_title, result_dict,
+                            language_title, lang_cld
+                            )
+                else:
+                    non_text += 1
+                    string_payload = '{0} {1}\n{2} {3}'.format(
+                        mime_title, fido.mime,
+                        puid_title, fido.puid
+                        )
+            else:
+                non_mime += 1
+                string_payload = ''
```

Comment:
Trying to remember the case of having no mime and if we want to be writing a WARC record at all in this case...

Reply:
Some of the urls from one WARC file that have no mime are:

http://www.whitehouse.gov.edgesuite.net/news/releases/2006/04/20060408.es.a.mp3
http://www.whitehouse.gov.edgesuite.net/news/radio/20040121-radioday-leavitt.v.smil
http://www.whitehouse.gov.edgesuite.net/omb/expectmore/css/print-sub.css
http://www.whitehouse.gov.edgesuite.net/omb/memoranda/fy04/m04-25_template.xls
http://www.whitehouse.gov/query.html?sc=0&qt=url%3Awww.whitehouse.gov/news/releases/2001/05&charset=iso-8859-1&ql=a&ct=1660165932&qm=1

Reply:
@ldko So, I played with fido on Friday. I ran through the different methods they have, which some don't work. In the warc-metadata-sidecar we are using the identify_stream method. That seems to be the best method to use to find mime and puid. In running that method for 5-6 different WARC files, it does not result in using 'extension'. When the string payload was empty it was because of 1 of 2 results, both in which did not find the mime type. Those are: `KO,3,,,,2526,"STDIN",,"fail"` and `OK,5,fmt/111,"OLE2 Compound Document Format","OLE2 Compound Document Format",104960,"STDIN","None","signature"`. The "None" is where the mime type should be on that second result.

Yesterday I played with python-magic and it does help fill in the blanks a bit, but I'm not sure if it's the results we want. For example... This, http://www.whitehouse.gov.edgesuite.net/omb/expectmore/css/screen-subType02.css, url's string payload would be empty when we run the script using identify_stream only. If we add in the extra check with python-magic this is our string payload:
>Identified-Payload-Type: text/plain
Preservation-Identifier: None
Charset-Detected: {'encoding': 'ascii', 'confidence': 1.0}
Languages-cld2: {'reliable': False, 'text-bytes': 1754, 'languages': [{'name': 'ENGLISH', 'code': 'en', 'text-covered': 18, 'score': 748.0}]}

I'm not sure that's what we should expect. If I add in a few extra steps to check the extension, using fido, then it does give us what we expect, but as @vphill  mentioned in slack about the extension not being accurate. I'm not sure what to do about the string payload.
Here is what I got from that same url when checking the extension:
>Identified-Payload-Type: text/css
Preservation-Identifier: x-fmt/224
Charset-Detected: {'encoding': 'ascii', 'confidence': 1.0}
Languages-cld2: {'reliable': False, 'text-bytes': 1754, 'languages': [{'name': 'ENGLISH', 'code': 'en', 'text-covered': 18, 'score': 748.0}]}